feat: Add V3 App Service Migration doc (#1068)
edgexfoundry/edgex-docs: docs_src/microservices/application/V3Migration.md

# V3 Migration Guide

## Configuration

The migration of any Application Service's configuration starts with migrating configuration common to all EdgeX services. See the [V3 Migration of Common Configuration](../../configuration/V3MigrationCommonConfig) section for details including the change from **TOML** format to **YAML** format for the configuration file. The remainder of this section focuses on configuration specific to Application Services.

### Common Configuration Removed

> All configuration that is common to all EdgeX services or all EdgeX Application Services needs to be removed from custom application service's private configuration. See [Common Service Configuration](../../configuration/CommonConfiguration/) section for details about configuration that is common to all Edgex services. See [Application Service Configuration](../GeneralAppServiceConfig) section for details about configuration that is common to all EdgeX Application Services.

Reviewer: maybe bullet this out to make it easier to read?

```suggestion
Any configuration common to all EdgeX services or all EdgeX Application Services needs to be removed from custom application service's private configuration. 
- See [Common Service Configuration](../../configuration/CommonConfiguration/) section for details about configuration that is common to all EdgeX services. 
- See [Application Service Configuration](../GeneralAppServiceConfig) section for details about configuration that is common to all EdgeX Application Services.
```

!!! note

>     With this change, the your custom application service must be run with either the `-cp--configProvider` flag or the `-cc/-commonConfig` flag in order for the service to receive the common configuration that has been removed from its private configuration. See [Config Provider](../../configuration/CommonCommandLineOptions/#config-provider) and [Common Config](../../configuration/CommonCommandLineOptions/#common-config) sections for more details on these flags.

Reviewer 2: should it be `--commonConfig` or `-commonConfig` ?

Reviewer 2: also, I think `-cp--configProvider` should be `-cp/--configProvider`

Author: fixed

>     With this change, the your custom application service must be run with either the `-cp/--configProvider` flag or the `-cc/--commonConfig` flag in order for the service to receive the common configuration that has been removed from its private configuration. See [Config Provider](../../configuration/CommonCommandLineOptions/#config-provider) and [Common Config](../../configuration/CommonCommandLineOptions/#common-config) sections for more details on these flags.

Reviewer: ```suggestion
    With this change, any custom application service must be run with either the `-cp/--configProvider` flag or the `-cc/--commonConfig` flag in order for the service to receive the common configuration that has been removed from its private configuration. See [Config Provider](../../configuration/CommonCommandLineOptions/#config-provider) and [Common Config](../../configuration/CommonCommandLineOptions/#common-config) sections for more details on these flags.
```

### MessageBus

> The EdgeX MessageBus configuration has been moved out of the Trigger configuration and most values are pulled from the common configuration. The only values that need to be in the application service's private configuration are:

Reviewer: ```suggestion
The EdgeX MessageBus configuration has been moved out of the Trigger configuration and most values are placed in the common configuration. The only values remaining in the application service's private configuration are:
```

- `Disabled` - Used to disable the use of the EdgeX MessageBus when not using metrics and not using `edgex-messagebus` Trigger type. Value need to be present so that it can be overridden with environment variable.
- `Optional.ClientId` - Unique name needed for when MQTT or NATS are used as the MessageBus implementation.

!!! example - "Example Application Service specific MessageBus section for 3.0"
   ```yaml
   MessageBus:
     Disabled: false  # Set to true if not using metrics and not using `edgex-messagebus` Trigger type
     Optional:

>        ClientId: "<service-key"

Author: fixed

   ```

### Trigger

#### edgex-messagebus changes

As noted above the EdgeX MessageBus configuration has been removed from the **Trigger** configuration. In addition the `SubscribeTopics` and `PublishTopic` settings have been move to the top level of the **Trigger** configuration. Most application services can simply use the default trigger configuration from application service common configuration.

!!! example - "Example application service Trigger configuration - From Common Configuration "
    ```yaml
    Trigger:
      Type: "edgex-messagebus"
      SubscribeTopics: "events/#" # Base topic is prepended to this topic when using edgex-messagebus
    ```

!!! example - "Example local application service Trigger configuration - **None**"
```yaml
# Using default Trigger config from common config
```

> Some application service many need to publish result back to the EdgeX MessageBus. In this case the `PublisTopic` will remain in the service private configuration.

Author: fixed

> Some application service many need to publish result back to the EdgeX MessageBus. In this case the `PublishTopic` will remain in the service private configuration.

Reviewer: ```suggestion
Some application services may need to publish results back to the EdgeX MessageBus. In this case, the `PublishTopic` will remain in the service private configuration.
```

!!! example - "Example local application service Trigger configuration - **PublishTopic**"
    ```yaml
    Trigger:
      # Default value for SubscribeTopics is aslo set in common config
      PublishTopic: "<my-topic>"  # Base topic is prepended to this topic when using edgex-messagebus
    ```

!!! note

>     In EdgeX 3.0 Application services, the base topic set in MessageBus common configuration is prepended to the configured `SubscribeTopics` and `PublishTopic` values. The default base topic is set to `edgex` so that all topics start with `edgex/`

Reviewer: ```suggestion
    In EdgeX 3.0 Application services, the base topic in MessageBus common configuration is prepended to the configured `SubscribeTopics` and `PublishTopic` values. The default base topic is `edgex`; thus, all topics start with `edgex/`.
```

        ClientId: "app-my-service"
        ConnectTimeout: "30s" 
        AutoReconnect: true
        KeepAlive: 10 # Seconds (must be 2 or greater)
        QoS: 0 # Quality of Service 0 (At most once), 1 (At least once) or 2 (Exactly once)
        Retain: true
        SkipCertVerify: false
        SecretName: "mqtt-trigger" 
        AuthMode: "none"
    ```

####  external-mqtt **Trigger** Migration

1. Move your `SubscribeTopics` and `PublishTopic` top level of the **Trigger** configuration

#### HTTP Changes

The HTTP trigger configuration has not changed for EdgeX 3.0

### Writable Pipeline

See [Pipeline Configuration](#pipeline-configuration) section below for changes to the Writable Pipeline configuration

## Custom Application Service 

### Code

#### Dependencies

You first need to update the `go.mod` file to specify `go 1.20` and the v3 versions of the App Functions SDK and any EdgeX go-mods directly used by your service. 

!!! example "Example go.mod for V3"

    ```go
    module <your service>
    
    go 1.20
    
    require (
    	github.com/edgexfoundry/app-functions-sdk-go/v3 v3.0.0
    	github.com/edgexfoundry/go-mod-core-contracts/v3 v3.0.0
    )
    ```

> 

Reviewer: Do we need to add a note about running `make tidy` or `go mod tidy` to update the rest of the go.mod/go.sum files? or is that implied?

### Pipeline Configuration

- **Writable.Pipeline.TargetType** has change from a bool to a string with valid values or `raw`, `event` or `metric`
- Topic wild cards have changed to conform 100% with MQTT scheme. The `#` between level has be replaced with `+` . See [Multi-level topics and wildcards](../../general/messagebus/#multi-level-topics-and-wildcards) for more details.
- **HTTPExport** function configuration
    - Parameter `SecretName`  renamed to be  `SecretValueKey`  
    - Parameter `SecretPath` renamed to be `SecretName` 
- **Encrypt** function configuration 
    - Parameter `SecretName`  renamed to be  `SecretValueKey`  
    - Parameter `SecretPath` renamed to be `SecretName` 

### Environment Variable Overrides

> Environment variable override must be adjusted appropriately for the above changes. Remove those overrides that apply to any common configuration.

Reviewer: ```suggestion
Environment variable overrides must be adjusted appropriately for the above changes. Remove any overrides that apply to common configuration.
```

### Common Configuration Removed

Any configuration that is common to all EdgeX services or all EdgeX Application Services needs to be removed from custom application service's private configuration. 

- See [Common Service Configuration](../../configuration/CommonConfiguration/) section for details about configuration that is common to all Edgex services. 
- See [Application Service Configuration](../GeneralAppServiceConfig) section for details about configuration that is common to all EdgeX Application Services.

!!! note

>     With this change, the any custom application service must be run with either the `-cp/--configProvider` flag or the `-cc/--commonConfig` flag in order for the service to receive the common configuration that has been removed from its private configuration. See [Config Provider](../../configuration/CommonCommandLineOptions/#config-provider) and [Common Config](../../configuration/CommonCommandLineOptions/#common-config) sections for more details on these flags.

Reviewer: take out `the` before `any` - so it should start `With this change, any...`
